hermes/probe/delete/delete.go

```diff
@@ -0,0 +1,133 @@
+func DeleteFile(ctx context.Context, fileID int32, target *probe.Target, client stiface.Client, logger *logger.Logger) (int32, error) {
+	bucket := target.Target.GetBucketName()
+
+	if fileID <= 10 || fileID >= 51 {
+		return fileID, fmt.Errorf("delete(%q, %q) failed; invalidArgument: expected fileID %d to be within valid inclusive range: 11-50", bucket, fileID, fileID)
+	}
+
+	// TODO(evanSpendlove): Add custom error object to return value and modify all returns.
+	filename, ok := target.Journal.Filenames[fileID]
+	if !ok {
+		return fileID, fmt.Errorf("StateJournalInconsistent: Journal.Filenames has no entry with file ID = %d", fileID)
+	}
+
+	target.Journal.Intent = &pb.Intent{
+		FileOperation: pb.Intent_CREATE,
+		Filename:      filename,
+	}
+
+	file := client.Bucket(bucket).Object(filename)
+
+	start := time.Now()
+	if err := file.Delete(ctx); err != nil {
+		var status metrics.ExitStatus
+		switch err {
+		case storage.ErrObjectNotExist:
+			status = metrics.FileMissing
+		case storage.ErrBucketNotExist:
+			status = metrics.BucketMissing
+		default:
+			status = metrics.ProbeFailed
+		}
+
+		target.LatencyMetrics.APICallLatency[metrics.APIDeleteFile][status].Metric(apiLatency).AddFloat64(time.Now().Sub(start).Seconds())
+		return fileID, fmt.Errorf("delete(%q, %q) failed; status %v: %w", bucket, filename, status, err)
+	}
+	target.LatencyMetrics.APICallLatency[metrics.APIDeleteFile][metrics.Success].Metric(apiLatency).AddFloat64(time.Now().Sub(start).Seconds())
+
+	query := &storage.Query{Prefix: filename}
+	start = time.Now()
+	objects := client.Bucket(bucket).Objects(ctx, query)
+	for {
+		obj, err := objects.Next()
+		if err == iterator.Done {
+			break
+		}
+		if obj.Name == filename {
+			status := metrics.ProbeFailed
+			target.LatencyMetrics.APICallLatency[metrics.APIListFiles][status].Metric(apiLatency).AddFloat64(time.Now().Sub(start).Seconds())
+			return fileID, fmt.Errorf("delete(%q, %q) failed; status %v: object %v still listed after delete", bucket, filename, status, obj.Name)
+		}
+		if err != nil {
+			status := metrics.BucketMissing
+			target.LatencyMetrics.APICallLatency[metrics.APIListFiles][status].Metric(apiLatency).AddFloat64(time.Now().Sub(start).Seconds())
+			return fileID, fmt.Errorf("delete(%q, %q) failed; status %v: %w", bucket, filename, status, err)
+		}
+	}
+	target.LatencyMetrics.APICallLatency[metrics.APIListFiles][metrics.Success].Metric(apiLatency).AddFloat64(time.Now().Sub(start).Seconds())
+
+	// Update in-memory NIL file after delete operation.
+	delete(target.Journal.Filenames, fileID)
+
+	logger.Infof("Object %v deleted in bucket %s.", file, bucket)
+	return fileID, nil
+}
+
+// PickFileToDelete picks which file to delete and returns the integer ID of this file.
+// Returns:
+//	- ID: returns the ID of the file to be deleted.
+func PickFileToDelete() int32 {
+	const (
+		begin                  = 11 // we can delete files starting from the file Hermes_11
```

Comment:
Maybe rename to `minFileIDToDelete`? Also, I thought that only files 1-9 were idempotent.

Reply:
I'll rename it accordingly. The nil file is file 0, so the first 10 files (1-10) are idempotent.